opennav_following/src/following_server.cpp

```diff
@@ -0,0 +1,595 @@
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+FollowingServer::on_activate(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Activating %s", get_name());
+
+  auto node = shared_from_this();
+
+  tf2_listener_ = std::make_unique<tf2_ros::TransformListener>(*tf2_buffer_);
+  vel_publisher_->on_activate();
+  filtered_dynamic_pose_pub_->on_activate();
+  following_action_server_->activate();
+
+  // Add callback for dynamic parameters
+  dyn_params_handler_ = node->add_on_set_parameters_callback(
+    std::bind(&FollowingServer::dynamicParametersCallback, this, _1));
+
+  // Create bond connection
+  createBond();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+FollowingServer::on_deactivate(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Deactivating %s", get_name());
+
+  following_action_server_->deactivate();
+  vel_publisher_->on_deactivate();
+  filtered_dynamic_pose_pub_->on_deactivate();
+
+  remove_on_set_parameters_callback(dyn_params_handler_.get());
+  dyn_params_handler_.reset();
+  tf2_listener_.reset();
+
+  // Destroy bond connection
+  destroyBond();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+FollowingServer::on_cleanup(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Cleaning up %s", get_name());
+  tf2_buffer_.reset();
+  following_action_server_.reset();
+  controller_.reset();
+  vel_publisher_.reset();
+  filtered_dynamic_pose_pub_.reset();
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+FollowingServer::on_shutdown(const rclcpp_lifecycle::State &)
+{
+  RCLCPP_INFO(get_logger(), "Shutting down %s", get_name());
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+template<typename ActionT>
+void FollowingServer::getPreemptedGoalIfRequested(
+  typename std::shared_ptr<const typename ActionT::Goal> goal,
+  const std::unique_ptr<nav2_util::SimpleActionServer<ActionT>> & action_server)
+{
+  if (action_server->is_preempt_requested()) {
+    goal = action_server->accept_pending_goal();
+  }
+}
+
+template<typename ActionT>
+bool FollowingServer::checkAndWarnIfCancelled(
+  std::unique_ptr<nav2_util::SimpleActionServer<ActionT>> & action_server,
+  const std::string & name)
+{
+  if (action_server->is_cancel_requested()) {
+    RCLCPP_WARN(get_logger(), "Goal was cancelled. Cancelling %s action", name.c_str());
+    return true;
+  }
+  return false;
+}
+
+template<typename ActionT>
+bool FollowingServer::checkAndWarnIfPreempted(
+  std::unique_ptr<nav2_util::SimpleActionServer<ActionT>> & action_server,
+  const std::string & name)
+{
+  if (action_server->is_preempt_requested()) {
+    RCLCPP_WARN(get_logger(), "Goal was preempted. Cancelling %s action", name.c_str());
+    return true;
+  }
+  return false;
+}
+
+void FollowingServer::followObject()
```

Comment:
It might be nice (if we change to use a topic instead of the preemption for following) to have the Request remove the Pose information and replace it with a TF2 frame. Maybe someone wants to track a frame rather than a pose topic. If a frame is given, use that. Else, use the subscription callback pose